### ISSUE TITLE
Errors when building platformio.ini file.

### ISSUE THREAD

**Issue by Purclewan** [jjssoftware/iD8266#7]
Hi Joe,
Thanks for your email.  I have been trying without success to use platformio as you suggested to build and flash your software to various ESP8266 boards.  Lastly to a Wemos D1.  When I try to run the platformio.ini file I get the following errors.  I know I am doing something stupid but just what I can't figure out what.  Appreciate you are not here to teach me but the the learning curve is steep and I am still finding my way round platformio so any pointers would be gratefully appreciated.  

Michael.

Linking .pioenvs\iD8266\firmware.elf
.pioenvs\iD8266\libFrameworkArduino.a(core_esp8266_main.cpp.o):(.text._ZL12loop_wrapperv+0x4): undefined reference to `setup'
.pioenvs\iD8266\libFrameworkArduino.a(core_esp8266_main.cpp.o):(.text._ZL12loop_wrapperv+0x8): undefined reference to `loop'
.pioenvs\iD8266\libFrameworkArduino.a(core_esp8266_main.cpp.o): In function `loop_wrapper()':
core_esp8266_main.cpp:(.text._ZL12loop_wrapperv+0x21): undefined reference to `setup'
core_esp8266_main.cpp:(.text._ZL12loop_wrapperv+0x2d): undefined reference to `loop'
collect2.exe: error: ld returned 1 exit status
*** [.pioenvs\iD8266\firmware.elf] Error 1
 [ERROR] Took 10.90 second

**Comment by bebo-dot-dev** [comment]
Hi Mike no problem, with a bit of luck we'll be able to workout what's going on here. 

Can you please tell me what operating system you use (i.e. Windows 7, Linux Ubuntu etc) and which version of platformIO you chose to install (i.e. Atom or VSCode plug in) ?

Let me know and I'll try and recreate the problem that you're seeing here.

**Comment by Purclewan** [comment]
Thanks Joe.
Running a fairly new build PC with new build Windows 7 64bit and 16G Ram Service Pack 1 (with all recent updates).  
Following are installed.
Java release 8.0.161 
Python 2.7.14
Platformio installed with Atom 1.25 
Nodejs 8.9.1
Eclipse Java Oxygen.
GitHub Desktop 1.10 
Arduino IDE 1.85 
Windows path variable, 
C:\Users\Michael\AppData\Local\Programs\Python\Python36\Scripts\;C:\Users\Michael\AppData\Local\Programs\Python\Python36\;C:\Users\Michael\AppData\Roaming\npm;C:\Users\Michael\AppData\Local\atom\bin;C:\Users\Michael\.platformio\penv\Scripts;C:\Users\Michael\AppData\Local\GitHubDesktop\bin

A little more background.  I have been using and coding with Arduino for several months but trying to move on.  I have successfully used platformio to flash project code from GitHub to Sonoff devices but admit I simply followed step by step instructions whilst trying to learn.  This is my first slightly more complex project.

I try not to pester developers with questions as I really appreciate the work they do demands a lot of time and to paraphrase Newton, without their efforts I would have no shoulders to stand on. However, this is the first time I have been beaten and I am totally at a loss so any pointer would be very welcome.

M.

**Comment by bebo-dot-dev** [comment]
Hi Mike, thanks for the detailed run down of your setup. 

I've not had any luck reproducing your problem here yet. I have grabbed a brand new IE11 Win7 x86 test VM from https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/ to test with. My steps:

1. Boot the new clean Windows 7 VM in VirtualBox
2. Install the latest nodeJS v8.10.0
3. Install all windows build tools and SDKs globally with `npm install --global --production windows-build-tools`
4. Install atom 1.25.0
5. Install the platformio package in atom
6. Download a clean copy of the Id8266 source as a zip from https://github.com/jjssoftware/iD8266/archive/master.zip and extract it
7. Open the project code in atom / platformio at the folder containing the platformio.ini file
8. Successfully build the code using the platformio build option in the left toolbar

There's a little video at the bottom of this message showing this working build. It compiled successfully first time here.

Unfortunately I can't get my hands on an x64 Windows 7 Virtual Machine to test with but I don't think this would change things. I have a feeling that your problem is environmental.

If I had to guess what the actual cause is, I would say that it's related to you having the Arduino IDE installed and Windows has associated the .ino file extension with the Arduino IDE on your system. If this theory holds up, this is causing the link stage to fail because earlier the iD8266.ino file wasn't found / included for compilation. 

Please try renaming the iD8266.ino file to iD8266.cpp in the src folder and then attempt another build in platformio to test this theory.

![20180321_211328](https://user-images.githubusercontent.com/536164/37738902-69468356-2d50-11e8-9204-3fb6f9c4e1bf.gif)

**Comment by Purclewan** [comment]
Joe, renaming the iD8266.ino file to iD8266.cpp seems to have done the trick.  I guess it must be an environment issue as you suggest.  Onwards and upwards:-)

**Comment by bebo-dot-dev** [comment]
Hi Mike, that's great news thanks for the update.

The only reason that the iD8266.ino file has a .ino file extension and not a .cpp file extension is to enable compilation / maintenance via the Arduino IDE. To be completely honest I'm not sure that this is still a requirement of the Arduino IDE. It would be nice if I could rename this file or find some other way to make sure that this doesn't happen again in the future. I'll run some tests.

**Comment by bebo-dot-dev** [comment]
Hi Mike (@Purclewan)

I've run some tests and come up with a solution. I hope this will the final update on this issue following commit: https://github.com/jjssoftware/iD8266/commit/80dfb93d44da5811765cabf25f1999a649369039.

I've restructured the code for Arduino IDE compatibility whilst at the same maintained compatibility with platformio. I've also solved the issue that you originally reported in this issue which was the source code in iD8266.ino not being picked up for compilation by platformio in your environment:

1. Code now resides in the iD8266 folder and includes an iD8266.ino file to enable the project to be opened and compiled in the Arduino IDE. This iD8266.ino file is simply a stub.
2. The iD8266 folder also includes an iD8266.cpp file that contains the setup and loop entry point functions required by the Arduino ESP8266 core.
3. platformio.ini now points to the iD8266 source folder.

I believe this will make the code compatible for your environment as-is, makes it generally more flexible and also reduces friction for new future users.
